casper/src/main/scala/coop/rchain/casper/dag/BlockDagKeyValueStorage.scala

```diff
@@ -327,4 +363,62 @@ object BlockDagKeyValueStorage {
       stores.deployPool,
       stores.invalidBlocks
     )
+
+  def deployStatus[F[_]: Sync](
+      d: DeployId,
+      deployStore: KeyValueTypedStore[F, DeployId, Signed[DeployData]],
+      deployIndex: KeyValueTypedStore[F, DeployId, BlockHash],
+      invalidBlocksIndex: KeyValueTypedStore[F, DeployId, BlockMetadata],
+      finalizedSet: Set[BlockHash],
+      executedMap: Map[DeployId, Option[String]],
+      expiredSet: Set[DeployId]
+  ): F[String] = {
+    val pooledF   = deployStore.collect { case (dId, _) => dId }.map(_.contains(d))
+    val expiredF  = Sync[F].delay(expiredSet.contains(d))
+    val includedF = deployIndex.collect { case (dId, _) => dId }.map(_.contains(d))
+    val orphanedF = invalidBlocksIndex.collect { case (dId, _) => dId }.map(_.contains(d))
+    val finalizedF = deployIndex.get1(d).map { hashOpt =>
+      hashOpt.exists(h => finalizedSet.contains(h))
+    }
+    val inProgress = if (executedMap.contains(d)) {
+      val done = executedMap(d).nonEmpty
+      val r =
+        if (done) s"Executed, result: ${executedMap(d).get}"
+        else "Execution is in progress."
+      r.some
+    } else none[String]
+
+    Stream(
+      inProgress.pure,
+      pooledF.map(v => v.guard[Option].as("Pooled")),
+      expiredF.map(v => v.guard[Option].as("Expired")),
+      orphanedF.map(v => v.guard[Option].as("Orphaned")),
+      includedF.map(v => v.guard[Option].as("Included")),
+      finalizedF.map(v => v.guard[Option].as("Finalized"))
+    ).covary[F]
```

Comment:
Could you explain this piece of code, please?

Reply:
There are several checks to be performed - each check returns whether it is found that deploy is in some status.
`pooledF` returns F[Boolean], 
`v.guard[Option].as("Pooled")` returns Some("Pooled") if v is true and None if v is false. 

`guard[F[_]]` is a syntax of an `Alternative` type class, and `cats` library has implementation of an `Alternative[Option]`

So each check returns a None or Some with deploy status string inside.

Then all these executions are packed as a Steam that represent a sequence of computations. What I want to do is evaluate all these computations (see `.map(Stream.eval)`) until first Some is found (see `.collectFirst { case Some(status) => status }`). This ensures that computation short circuits, so Stream execution stops and if `inProgress.pure` returns `Some(something)` nothing else is executed.